aws/resource_aws_gamelift_fleet.go

```diff
@@ -0,0 +1,596 @@
+						},
+					},
+				},
+			},
+			"runtime_configuration": {
+				Type:     schema.TypeList,
+				MaxItems: 1,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"game_session_activation_timeout_seconds": {
+							Type:         schema.TypeInt,
+							Optional:     true,
+							ValidateFunc: validation.IntBetween(1, 600),
+						},
+						"max_concurrent_game_session_activations": {
+							Type:         schema.TypeInt,
+							Optional:     true,
+							ValidateFunc: validation.IntBetween(1, 2147483647),
+						},
+						"server_process": {
+							Type:     schema.TypeList,
+							Optional: true,
+							MaxItems: 50,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"concurrent_executions": {
+										Type:         schema.TypeInt,
+										Required:     true,
+										ValidateFunc: validation.IntAtLeast(1),
+									},
+									"launch_path": {
+										Type:         schema.TypeString,
+										Required:     true,
+										ValidateFunc: validation.StringLenBetween(1, 1024),
+									},
+									"parameters": {
+										Type:         schema.TypeString,
+										Optional:     true,
+										ValidateFunc: validation.StringLenBetween(1, 1024),
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceAwsGameliftFleetCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).gameliftconn
+
+	input := gamelift.CreateFleetInput{
+		BuildId:         aws.String(d.Get("build_id").(string)),
+		EC2InstanceType: aws.String(d.Get("ec2_instance_type").(string)),
+		Name:            aws.String(d.Get("name").(string)),
+	}
+
+	if v, ok := d.GetOk("description"); ok {
+		input.Description = aws.String(v.(string))
+	}
+	if v, ok := d.GetOk("ec2_inbound_permission"); ok {
+		input.EC2InboundPermissions = expandGameliftIpPermissions(v.([]interface{}))
+	}
+	if v, ok := d.GetOk("metric_groups"); ok {
+		input.MetricGroups = expandStringList(v.([]interface{}))
+	}
+	if v, ok := d.GetOk("new_game_session_protection_policy"); ok {
+		input.NewGameSessionProtectionPolicy = aws.String(v.(string))
+	}
+	if v, ok := d.GetOk("resource_creation_limit_policy"); ok {
+		input.ResourceCreationLimitPolicy = expandGameliftResourceCreationLimitPolicy(v.([]interface{}))
+	}
+	if v, ok := d.GetOk("runtime_configuration"); ok {
+		input.RuntimeConfiguration = expandGameliftRuntimeConfiguration(v.([]interface{}))
+	}
+
+	log.Printf("[INFO] Creating Gamelift Fleet: %s", input)
+	out, err := conn.CreateFleet(&input)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(*out.FleetAttributes.FleetId)
+
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{
+			gamelift.FleetStatusActivating,
+			gamelift.FleetStatusBuilding,
+			gamelift.FleetStatusDownloading,
+			gamelift.FleetStatusNew,
+			gamelift.FleetStatusValidating,
+		},
+		Target:  []string{gamelift.FleetStatusActive},
```

Comment:
So I think we provide as much details as we can in such case. I'm intentionally _not_ catching `FleetStatusError` and let it fall through as error, so I can catch it below and ask for reasons of the state.

We also save some API calls that way since we don't ask for failures with each retry, but only when we _know_ things have failed.

Reply:
Okay that makes sense, and is pretty awesome. 👍